From 013ace76f054db714d74c762362896e292cfec32 Mon Sep 17 00:00:00 2001
From: Sugar Zhang <sugar.zhang@rock-chips.com>
Date: Wed, 28 Nov 2018 20:10:53 +0800
Subject: [PATCH] ASoC: rockchip: i2s_tdm: Add reset mechanism

This patch brings i2s back to normal by resetting i2s tx/rx
relative logic if i2s' clear operation is failed.

Change-Id: I52e4713d26f781962278802bd1f9bbce3fe4b751
Signed-off-by: Sugar Zhang <sugar.zhang@rock-chips.com>
---
 sound/soc/rockchip/rockchip_i2s_tdm.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/sound/soc/rockchip/rockchip_i2s_tdm.c b/sound/soc/rockchip/rockchip_i2s_tdm.c
index 8272a5853a2b..35199e843dac 100644
--- a/sound/soc/rockchip/rockchip_i2s_tdm.c
+++ b/sound/soc/rockchip/rockchip_i2s_tdm.c
@@ -308,7 +308,8 @@ static void rockchip_snd_txrxctrl(struct snd_pcm_substream *substream,
 				regmap_read(i2s_tdm->regmap, I2S_CLR, &val);
 				retry--;
 				if (!retry) {
-					dev_info(i2s_tdm->dev, "fail to clear\n");
+					dev_info(i2s_tdm->dev, "reset txrx\n");
+					rockchip_snd_xfer_sync_reset(i2s_tdm);
 					break;
 				}
 			}
@@ -349,7 +350,10 @@ static void rockchip_snd_txctrl(struct rk_i2s_tdm_dev *i2s_tdm, int on)
 			regmap_read(i2s_tdm->regmap, I2S_CLR, &val);
 			retry--;
 			if (!retry) {
-				dev_warn(i2s_tdm->dev, "fail to clear\n");
+				dev_warn(i2s_tdm->dev, "reset tx\n");
+				reset_control_assert(i2s_tdm->tx_reset);
+				udelay(1);
+				reset_control_deassert(i2s_tdm->tx_reset);
 				break;
 			}
 		}
@@ -388,7 +392,10 @@ static void rockchip_snd_rxctrl(struct rk_i2s_tdm_dev *i2s_tdm, int on)
 			regmap_read(i2s_tdm->regmap, I2S_CLR, &val);
 			retry--;
 			if (!retry) {
-				dev_warn(i2s_tdm->dev, "fail to clear\n");
+				dev_warn(i2s_tdm->dev, "reset rx\n");
+				reset_control_assert(i2s_tdm->rx_reset);
+				udelay(1);
+				reset_control_deassert(i2s_tdm->rx_reset);
 				break;
 			}
 		}
@@ -568,7 +575,8 @@ static void rockchip_i2s_tdm_xfer_pause(struct snd_pcm_substream *substream,
 		regmap_read(i2s_tdm->regmap, I2S_CLR, &val);
 		retry--;
 		if (!retry) {
-			dev_info(i2s_tdm->dev, "fail to clear\n");
+			dev_info(i2s_tdm->dev, "reset txrx\n");
+			rockchip_snd_xfer_sync_reset(i2s_tdm);
 			break;
 		}
 	}
-- 
2.35.3

